Wien2k: core-hole parsing

jfrodlova opened the pull request.

ondracka commented:
Hi @jfrodlova , this looks really good. I still need to reread the Wien2k docs to make 100% sure all is OK, but at first glance this looks spot on. We might need to add few more quantities, but right now the inc parsing itself
seems complete. Attached is a testcase for this, can you hook it up? 

[TiN-corehole.zip](https://github.com/user-attachments/files/16675118/TiN-corehole.zip)

ondracka commented:
Sorry about the force push to develop, I have some student here also doing a bit of nomad work as well and he was too hasty with a merge button (so I needed to clean it up afterwards). To clean up the mess it caused here, you should probably just backup the c6ae7fe626523d1dc4f5cf04a84b3fcb8c2236d6 somewhere, reset [jfrodlova:wien2k_core_hole](https://github.com/jfrodlova/electronic-parsers/tree/wien2k_core_hole) to ondracka/develop and cherry-pick c6ae7fe626523d1dc4f5cf04a84b3fcb8c2236d6 back?

ondracka commented on electronicparsers/wien2k/parser.py:
```diff
@@ -1053,7 +1097,40 @@ def parse_method(self):
+            self.inc_parser.parse()
+            atoms = self.inc_parser.get('one_atom')
+            for atom_index, atom in enumerate(atoms):
+                kappas = atom.get('kappas')
```
Could you please add link in the comments to the userguide http://www.wien2k.at/reg_user/textbooks/usersguide.pdf and also mention the specific Table 6.6: Relativistic quantum numbers ?

ondracka commented on electronicparsers/wien2k/parser.py:
```diff
@@ -1011,6 +1050,11 @@ def parse_method(self):
+        # find inc file and set as a mainfile for Wien2kParser
+        inc_file = self.get_wien2k_file('inc')
+        self.inc_parser.mainfile = inc_file
+        # TODO: handle when inc_file is None
```
Is the TODO still relevant? I see that the whole corehole parsing is inside `if self.inc_parser.mainfile:`.

jfrodlova replied:
Right, thanks for pointing that out.

jfrodlova replied:
Corrected.

ondracka commented on tests/data/wien2k/core_hole/TiN-corehole.inc:
```diff
@@ -0,0 +1,29 @@
+ 4 0.00  0  NUMBER OF ORBITALS (EXCLUDING SPIN), SHIFT, IPRINT
```
Did I provide just the scf and inc file (I don't remember now and can§t find the email)? If you have the full directory, commiting all of it to the tests would be preferable.

ondracka replied:
Also please make the directory structure tests/data/wien2k/TiN-corehole/TiN-corehole.*
the Wien2k standard (for whatever historical reasons) is that the directory is named the same as the files inside...

jfrodlova replied:
Okay, I added the rest of TiN-corehole files to the directory and renamed it.

ondracka commented on tests/test_wien2kparser.py:
```diff
@@ -121,3 +124,15 @@ def test_dos(parser, caplog):
+    atom_par = sec_method.atom_parameters[0]
+    assert atom_par.atom_index == 2
+    assert atom_par.core_hole.n_electrons_excited == 1.0
+    assert atom_par.core_hole.l_quantum_number == 1.0
```
Nitpick: l_quantum_number should be integer...

ondracka commented on tests/test_wien2kparser.py:
```diff
@@ -121,3 +124,15 @@ def test_dos(parser, caplog):
+    assert atom_par.core_hole.n_electrons_excited == 1.0
+    assert atom_par.core_hole.l_quantum_number == 1.0
+    assert atom_par.core_hole.n_quantum_number == 2
+    assert atom_par.core_hole.occupation == 3
```
Please add test also for j_quantum_number

ondracka commented on electronicparsers/wien2k/parser.py:
```diff
@@ -1053,7 +1097,40 @@ def parse_method(self):
+                        core_hole.n_quantum_number = n_quantum_number
+                        core_hole.n_electrons_excited = electrons_excited
+                        core_hole.occupation = occupancy
+                        atom_index_arch = atom_par.append(atom_obj)
```
Do we use  atom_index_arch anywhere?

jfrodlova replied:
Corrected.

ondracka commented on electronicparsers/wien2k/parser.py:
```diff
@@ -1053,7 +1097,40 @@ def parse_method(self):
+                            spin_quantum_number = -1
+                        else:
+                            spin_quantum_number = 1
+                            j_quantum_number = (
+                                -kappas[orbital] / spin_quantum_number - 1 / 2
+                            )
+                        l_quantum_number = j_quantum_number - (spin_quantum_number / 2)
```
Nice :smiley:, I would have probably just hardcoded the table to a dict, but this is sure more concise, and if we add link to the table, it should be fine... Just make sure to comment that this calculations are based on the table.

ondracka reviewed (commented):
Thanks @jfrodlova This is great, please clean up the minor comments, make a PR upstream, and CC me there.

jfrodlova committed:
Wien2k: core-hole parsing

ondracka commented on tests/test_wien2kparser.py:
```diff
@@ -121,3 +124,16 @@ def test_dos(parser, caplog):
+    sec_method = archive.run[0].method[0]
+    atom_par = sec_method.atom_parameters[0]
+    assert atom_par.atom_index == 2
+    assert atom_par.core_hole.n_electrons_excited == 1
```
This one is actually a float (although this is kinda not really physically intuitive to speak about half-electron :smile: )

jfrodlova replied:
I agree with you but this is actually in definition, too. 

https://nomad-lab.eu/prod/v1/gui/analyze/metainfo/nomad/section_definitions@nomad.datamodel.metainfo.simulation.method.AtomParameters/core_hole/n_electrons_excited

Should I change the definition of n_electrons_excited?

ondracka replied:
This is what I get when I follow your link... 
![image](https://github.com/user-attachments/assets/25cd9898-eb18-4e83-a340-bcfaa63c392f)

`type: float64 (numpy)` but maybe you are refering to something else?

jfrodlova replied:
> This one is actually a float (although this is kinda not really physically intuitive to speak about half-electron 😄 )

I am not sure what you mean. Would you like to change the assert like this: 
`assert atom_par.core_hole.n_electrons_excited == 1.0`?

ondracka replied:
Yeah, that would be way better, but I guess this is probably safe in python as well :shrug:

ondracka replied:
Anyway this is nitpicking at this point...

jfrodlova replied:
No problem. Done.

ondracka commented on electronicparsers/wien2k/parser.py:
```diff
@@ -1053,7 +1096,44 @@ def parse_method(self):
+                        atom_obj.atom_index = atom_index
+                        core_hole = CoreHole()
+                        atom_obj.core_hole = core_hole
+                        j_quantum_number = np.array([j_quantum_number])
```
Do we need this line (the np.array particularly)? Later we test just `assert atom_par.core_hole.j_quantum_number == 1.5` which I believe is coorect...

jfrodlova replied:
In the definition of j_quantum_number (see the link below), I found that shape should be [1..2], so I created an array. (Otherwise, tests failed on that.)  

https://nomad-lab.eu/prod/v1/gui/analyze/metainfo/nomad/section_definitions@nomad.datamodel.metainfo.simulation.method.AtomParameters/core_hole/j_quantum_number

ondracka replied:
OK, fair enough

ondracka commented:
Great @jfrodlova I spotted two more minor issues, but with that fixed, you can just directly create a PR upstream.

jfrodlova committed:
Address PR issues

- delete unnecessary files from TiN-corehole test suite
- add logger for missing inc file
- add dscf state to CoreHole() and test it
